sklearn/impute.py

```diff
@@ -362,6 +379,26 @@ def _dense_fit(self, X, strategy, missing_values, fill_value):
         # Constant
         elif strategy == "constant":
             return np.full(X.shape[1], fill_value, dtype=X.dtype)
+    
+        # Sample
+        elif strategy == "sample":
+            X = X.transpose()
+            mask = mask.transpose()
+
+            generators = np.empty(X.shape[0], dtype=object)
+
+            for i, (row, row_mask) in enumerate(zip(X[:], mask[:])):
+                row_mask = np.logical_not(row_mask).astype(np.bool)
+                row = row[row_mask]
+                if row.size > 0:
+                    uniques, counts = np.unique(row, return_counts=True)
```

Comment:
if `all(counts == 1)`, which is likely for float valued features, we could just set proba = None.

Reply:
I'm also not entirely sure we should be doing unique, rather than just storing all the values from X, including repetitions.

Reply:
I agree that for inputs with float dtypes, it's very likely that ``all(counts == 1)``. However, for integer dtypes or categorical inputs (which this strategy allows), it's very likely that storing uniques and counts would be way less memory consuming.
I think the best is to not use uniques for float and use unique + counts otherwise.
What do you think ?

Reply:
I'd be happy with always using unique + counts, but setting p=None when len(uniques) == n_non_missing)